### PR TITLE
Simplex cc integration

### PR DESCRIPTION
**Pull request by mrose17**
if `runtime.config.simplex` is defined via the usual configuration mechanism, then information necessary for a `POST` to purchase BTC  for the browser's BAT wallet is returned on `GET /v2/wallet/{paymentId}`

### CODE REVIEW

File: bat-utils/lib/runtime-wallet.js [hunk 1]
```diff
@@ -475,4 +496,118 @@ Wallet.providers.mock = {
...
+// courtesy of https://stackoverflow.com/questions/2076299/how-to-close-current-tab-in-a-browser-window#18776480
+// javascript:window.open('','_self').close();
+// courtesy of http://www.yournewdesigner.com/css-experiments/javascript-window-close-firefox.html
+// javascript:window.open('','_parent','').close();
+
+    return ({
+      extend: { simplex: quote },
+      buyForm: {
+        method: 'POST',
+        action: this.runtime.config.simplex.url + '/payments/new',
```
Comment (evq) [inline bat-utils/lib/runtime-wallet.js]: updating with the results on our out of band discussion:
I would prefer that we did not make this url server provided but it was discussed and we have decided to proceed as implemented, we will enforce that the url provided is https in the browser

Comment (mrose17) [reply bat-utils/lib/runtime-wallet.js]: i would be happy to add a signature over the URL provided by a third-party that the browser could verify in a subsequent revision.

File: bat-utils/lib/runtime-wallet.js [hunk 2]
```diff
@@ -475,4 +496,118 @@ Wallet.providers.mock = {
...
+        authorization: 'ApiKey ' + this.runtime.config.simplex.api_key,
+        'content-type': 'application/json'
+      },
+      payload: JSON.stringify({
+        account_details: {
+          app_provider_id: params.partner,
+          app_version_id: params.version,
+          app_end_user_id: quote.user_id,
+          signup_login: {
+            ip: '213.162.55.5',
```
Comment (evq) [inline bat-utils/lib/runtime-wallet.js]: what is the significance of this ip?

File: bat-utils/lib/runtime-wallet.js [hunk 3]
```diff
@@ -475,4 +496,118 @@ Wallet.providers.mock = {
...
+              currency: params.currency,
+              address: info.addresses.BTC
+            }
+          }
+        }
+      })
+    })
+    if (Buffer.isBuffer(result)) try { result = JSON.parse(result) } catch (ex) { result = result.toString() }
+
+// courtesy of https://stackoverflow.com/questions/2076299/how-to-close-current-tab-in-a-browser-window#18776480
```
Comment (evq) [inline bat-utils/lib/runtime-wallet.js]: can we remove these comments? they don't seem related

File: bat-utils/lib/runtime-wallet.js [hunk 4]
```diff
@@ -475,4 +496,118 @@ Wallet.providers.mock = {
...
+    quote = info.simplex
+    if ((!quote) || (quote.expires <= now)) {
+      result = await braveHapi.wreck.post(this.runtime.config.simplex.url + '/wallet/merchant/v2/quote', {
+        headers: {
+          accept: 'application/json',
+          authorization: 'ApiKey ' + this.runtime.config.simplex.api_key,
+          'content-type': 'application/json'
+        },
+        payload: JSON.stringify({
+          end_user_id: info.paymentId,
```
Comment (evq) [inline bat-utils/lib/runtime-wallet.js]: is this the user wallet's paymentId? do we really need to send that to simplex? can we deterministicly generate an id only used with simplex?

File: ledger/controllers/wallet.js [hunk 5]
```diff
@@ -110,9 +111,23 @@ const read = function (runtime, apiVersion) {
               $set: { unsignedTx: result.unsignedTx }
             }
           }
-          await wallets.update({ paymentId: paymentId }, state, { upsert: true })
         }
       }
+
```
Comment (evq) [inline ledger/controllers/wallet.js]: I would prefer we set `state=null` here rather then clear after use above

File: bat-utils/lib/runtime-wallet.js [hunk 6]
```diff
@@ -475,4 +496,118 @@ Wallet.providers.mock = {
...
+    })
+    if (Buffer.isBuffer(result)) try { result = JSON.parse(result) } catch (ex) { result = result.toString() }
+
+// courtesy of https://stackoverflow.com/questions/2076299/how-to-close-current-tab-in-a-browser-window#18776480
+// javascript:window.open('','_self').close();
+// courtesy of http://www.yournewdesigner.com/css-experiments/javascript-window-close-firefox.html
+// javascript:window.open('','_parent','').close();
+
+    return ({
+      extend: { simplex: quote },
```
Comment (evq) [inline bat-utils/lib/runtime-wallet.js]: I would prefer a more descriptive name than extend. How about `walletQuote`?

File: bat-utils/lib/runtime-wallet.js (hunk 2 above)
Comment (mrose17) [reply bat-utils/lib/runtime-wallet.js]: the significance is that it is a non-existent IP address, cf., https://tools.ietf.org/html/rfc1918

we aren't going to supply the browser's IP address to the provider.

File: bat-utils/lib/runtime-wallet.js (hunk 3 above)
Comment (mrose17) [reply bat-utils/lib/runtime-wallet.js]: i plan to after the browser-side implementation is done. these links are the result of "considerable" research in how to use javascript to close the current tab. i thought it would be as simple as `_self.close()`, but nothing is that simple, sadly.

File: bat-utils/lib/runtime-wallet.js (hunk 4 above)
Comment (mrose17) [reply bat-utils/lib/runtime-wallet.js]: we can certainly add another stable UUID. i will do that.

File: bat-utils/lib/runtime-wallet.js (hunk 6 above)
Comment (mrose17) [reply bat-utils/lib/runtime-wallet.js]: ok.

File: ledger/controllers/wallet.js (hunk 5 above)
Comment (mrose17) [reply ledger/controllers/wallet.js]: why? the logical place to put it is after it is used, so if it isn't set again, we know we don't have to save it?

File: bat-utils/lib/runtime-wallet.js (hunk 3 above)
Comment (evq) [reply bat-utils/lib/runtime-wallet.js]: there are other mediums more suited for sharing and storing this research

File: ledger/controllers/wallet.js (hunk 5 above)
Comment (evq) [reply ledger/controllers/wallet.js]: my thinking is that clearing it right before use is defensive since it makes it less likely that we add an additional branch that doesn't clear it

File: bat-utils/lib/runtime-wallet.js (hunk 2 above)
Comment (evq) [reply bat-utils/lib/runtime-wallet.js]: sure, the browser will do that itself when it follows the link to go to the provider page :p 

I was wondering in the context of simplex but based on your answer it's the expected ip of the user signing up and can be wrong / spoofed without issue.
